Se tomó una versión reducida de **Million Songs Dataset** que mide 1.8 Gb, y se divide en dos archivos csv: *triplet_file.csv* y *metadata_file*.csv, en total el primer archivo cuenta con **2M** de registros y el segundo con **1M**, esto en la primera revisión y sin la limpieza de datos duplicados.

***Su formato original con un millón de canciones contemporáneas y populares, la cual mide un total de 493 Gb de los cuales quedan 272 Gb.***





In [ ]:
import pandas as pd
import numpy as np
#import Recommenders as Recommenders

Nos traemos la libreria del drive para poder acceder a los dos archivos.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Primer archivo (**1Gb**)

Nombre: Song Data

In [ ]:
song_df_1 = pd.read_csv('/content/drive/MyDrive/API Google/song_data.csv')
print(len(song_df_1))


1000000


Revisamos cuantos songid son unicos y cuantos son duplicados

In [ ]:
len(song_df_1.song_id.unique())

999056

In [ ]:
h = song_df_1.song_id.duplicated()
song_df_1[h]


song_id                      title  \
18309   SOPFBUZ12AF729C896             Spitting Games   
32760   SOJHKXX12A8C139798  A Crow Left Of The Murder   
33033   SOLTAEJ12A8C13F793              What If I Do?   
49718   SONEJDQ12A6D4FA71C          Starfighter Pilot   
54218   SONVJUL12A6701FB7A   The Damage In Your Heart   
...                    ...                        ...   
997476  SOCGXXL12B0B808865                 Yesterdays   
998785  SOEUIVS12A8C1400E3  Someone Put Your Hand Out   
999365  SORANLE12A6310DBC8             Long Way To Go   
999643  SOAPUFN12AB017FF03                Suggestions   
999740  SODNVTM12A6D4FA448           Ruled By Secrecy   

                              release                artist_name  year  
18309                     Final Straw                Snow Patrol  2003  
32760           Live in Malaysia 2004                    Incubus  2004  
33033                   In Your Honor               Foo Fighters  2005  
49718            Songs For Polarbears                Snow Patrol  1998  
54218                    Make Believe                     Weezer  2005  
...                               ...                        ...   ...  
997476                  Greatest Hits              Guns N' Roses  1991  
998785                    King Of Pop            Michael Jackson  1992  
999365          Love Angel Music Baby  Gwen Stefani / Andre 3000     0  
999643  System Of A Down (Bonus Pack)           System of a Down  1998  
999740                     Absolution                       Muse     0  

[944 rows x 5 columns]

999056 son unicos y 944 registros estan duplicados.

Ejemplo de la song id **SODNVTM12A6D4FA448**	

In [ ]:
song_df_1[song_df_1["song_id"] == "SODNVTM12A6D4FA448"]

song_id             title     release artist_name  year
766340  SODNVTM12A6D4FA448  Ruled By Secrecy  Absolution        Muse     0
999740  SODNVTM12A6D4FA448  Ruled By Secrecy  Absolution        Muse     0

Revisemos por el titulo de la cancion

In [ ]:
len(song_df_1.title.unique())


702429

In [ ]:
f = song_df_1.title.duplicated()
song_df_1[f]

song_id                title  \
101     SOPRQGM12A58A784F1                 Ride   
657     SOPILLI12A6D4FB80D                Intro   
806     SOGLLLR12A8AE470DE  Dancing In The Dark   
1180    SOTADIJ12AB0183DA2                Rowdy   
1224    SOVHTPX12A8C140778                Intro   
...                    ...                  ...   
999988  SOYKHSL12AF72A35AC     Wouldn't Get Far   
999989  SOIDAMG12AB0182B3A                Chill   
999990  SOFAXZU12AB01881C4             Way Down   
999991  SOYFFWW12A6701C135       So Many People   
999998  SOLXGOR12A81C21EB7              Faraday   

                                 release  \
101                                 Ride   
657                                 Heat   
806                  Dancing In The Dark   
1180                     Sunset...the EP   
1224                             Kinfolk   
...                                  ...   
999988                  Wouldn't Get Far   
999989  Oxygen Refreshed (Instrumentals)   
999990                          Way Down   
999991                              Four   
999998      The Trance Collection Vol. 2   

                                 artist_name  year  
101                                   Tiësto  2006  
657                                  H.e.a.t  2008  
806                    Micky Modelle v Jessy     0  
1180                            Tanya Morgan     0  
1224                   Ali & Gipp / Big Rube  2002  
...                                      ...   ...  
999988                              The Game     0  
999989                           Jazz Addixx     0  
999990  Taylor Hawkins & The Coattail Riders  2010  
999991                             Bah Samba  2002  
999998                                 Elude     0  

[297571 rows x 5 columns]

In [ ]:
song_df_1[song_df_1["title"] == "Dancing In The Dark"]

song_id                title  \
53      SOHMYGC12A6D4FAC4B  Dancing In The Dark   
806     SOGLLLR12A8AE470DE  Dancing In The Dark   
44158   SOCGTLB12AC468BDB4  Dancing In The Dark   
68697   SOIJJZS12A67ADC202  Dancing In The Dark   
95670   SOFWLYM12A58A7F45C  Dancing In The Dark   
106633  SOFVRCL12AB018377D  Dancing In The Dark   
178978  SOICIWS12A8C13A419  Dancing In The Dark   
368356  SOTRXIY12A8C13CDEA  Dancing In The Dark   
388942  SOCJKJX12A58A7DF6D  Dancing In The Dark   
399816  SOPKCJE12A8C140052  Dancing In The Dark   
401506  SOMHMUZ12A6D4F9210  Dancing In The Dark   
543049  SONJUPA12A8C146B05  Dancing In The Dark   
609708  SOLWNVX12AB01846B9  Dancing In The Dark   
700345  SOJPTKM12A6D4F9131  Dancing In The Dark   
767651  SOXIZLJ12A58A76655  Dancing In The Dark   
794605  SOMYVLD12A8C144333  Dancing In The Dark   
824571  SOARZIN12A8AE470D7  Dancing In The Dark   
931726  SOZJQSR12AB0189769  Dancing In The Dark   
940372  SOSQHMD12A58A7D171  Dancing In The Dark   
944178  SODSCIT12A8C14160E  Dancing In The Dark   
951971  SODXNWI12AB0182B5A  Dancing In The Dark   

                                       release  \
53                            The Look Of Love   
806                        Dancing In The Dark   
44158                          At Your Request   
68697                               Visitation   
95670                        Ellington Indigos   
106633                     They Sold A Million   
178978  You Broke My Heart In Seventeen Places   
368356  You Broke My Heart In Seventeen Places   
388942       Jo Stafford Sings Broadway's Best   
399816             Disco Discharge. Euro Disco   
401506                               Standards   
543049                     Dancing In The Dark   
609708         Esquivel - Orchestra and Chorus   
700345                     Dancing In The Dark   
767651                     Dancing In The Dark   
794605               This One's From The Heart   
824571                     Dancing In The Dark   
931726                       On The Sunny Side   
940372     Come Fly With Me - 18 Golden Greats   
944178                   Treasury Shows Vol. 9   
951971                      Sings Showstoppers   

                                   artist_name  year  
53                                 Diana Krall  2001  
806                      Micky Modelle v Jessy     0  
44158                              Cliff Adams     0  
68697                            Pekka Pohjola  1979  
95670         Duke Ellington And His Orchestra  2003  
106633              Artie Shaw & His Orchestra     0  
178978                           Tracey Ullman  1992  
368356                           Tracey Ullman  1992  
388942                             Jo Stafford     0  
399816                             Mike Mareen  1986  
401506                             Sonny Clark  1998  
543049                           Sonny Rollins  1987  
609708                                Esquivel  1996  
700345                   Micky Modelle v Jessy     0  
767651  Ray Anthony Orchestra - Big Band Swing  1995  
794605                            James Darren  1999  
824571                   Micky Modelle v Jessy     0  
931726                           The Four Lads     0  
940372                           Frank Sinatra  1958  
944178        Duke Ellington And His Orchestra  2003  
951971                             Jane Morgan     0

A pesar que la de Dancing In the Dark esta repetida varias veces, no significa que es la misma cancion ni ka misma version o autor

In [ ]:
len(song_df_1.artist_name.unique())

72665

72,665 Nombres de cantante unico

In [ ]:
song_df_1.year.unique()

array([2003, 1995, 2006,    0, 1993, 2007, 1997, 2005, 1998, 1994, 2009,
       1996, 2001, 2004, 2002, 1977, 2008, 1992, 1979, 1988, 2010, 1990,
       1981, 2000, 1978, 1991, 1989, 1999, 1975, 1969, 1982, 1966, 1972,
       1967, 1984, 1950, 1986, 1970, 1959, 1971, 1980, 1987, 1985, 1963,
       1973, 1958, 1961, 1938, 1939, 1983, 1929, 1964, 1948, 1968, 1976,
       1965, 1960, 1974, 1945, 1962, 1957, 1956, 1954, 1943, 1952, 1953,
       1955, 1951, 1931, 1928, 1937, 1949, 1935, 1941, 1947, 1934, 1940,
       1926, 1932, 1922, 1930, 1936, 1942, 1944, 1933, 1946, 1927, 2011,
       1925, 1924])

Segundo archivo (2Gb)

Nombre: Triplets File

In [ ]:
song_df_2 = pd.read_csv('/content/drive/MyDrive/API Google/triplets_file.csv')
song_df_2.head()
print(len(song_df_2))


2000000


Este archivo tiene la lista del conteo de canciones escuchadas y el id del usuario que las escucho


In [ ]:
song_df_2

user_id             song_id  \
0        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOAKIMP12A8C130995   
1        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBBMDR12A8C13253B   
2        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBXHDL12A81C204C0   
3        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBYHAJ12A6701BF1D   
4        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SODACBL12A8C13C273   
...                                           ...                 ...   
1999995  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJEYPO12AAA8C6B0E   
1999996  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJJYDE12AF729FC16   
1999997  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJKQSF12A6D4F5EE9   
1999998  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJUXGA12AC961885C   
1999999  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJYOLS12A8C13C06F   

         listen_count  
0                   1  
1                   2  
2                   1  
3                   1  
4                   1  
...               ...  
1999995             2  
1999996             4  
1999997             3  
1999998             1  
1999999             1  

[2000000 rows x 3 columns]

In [ ]:
len(song_df_2.user_id.unique())

76353

Encontramos 76 353 id usuarios unicos

In [ ]:
len(song_df_2.song_id.unique())

10000

10,000 id de canciones unicas escuchadas

In [ ]:
song_df_2.listen_count.max()

2213

In [ ]:
#Combinando información
song_df = pd.merge(song_df_2, song_df_1.drop_duplicates(['song_id']), on='song_id', how='left')
song_df.head()


user_id             song_id  listen_count  \
0  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOAKIMP12A8C130995             1   
1  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBBMDR12A8C13253B             2   
2  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBXHDL12A81C204C0             1   
3  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBYHAJ12A6701BF1D             1   
4  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SODACBL12A8C13C273             1   

             title                        release    artist_name  year  
0         The Cove             Thicker Than Water   Jack Johnson     0  
1  Entre Dos Aguas            Flamenco Para Niños  Paco De Lucia  1976  
2         Stronger                     Graduation     Kanye West  2007  
3   Constellations              In Between Dreams   Jack Johnson  2005  
4     Learn To Fly  There Is Nothing Left To Lose   Foo Fighters  1999

In [ ]:
len(song_df)

2000000

In [ ]:
print(len(song_df_1), len(song_df_2))

1000000 2000000


Se crea un nueva llave con la combinacion de titulo y artista

In [ ]:

song_df['song'] = song_df['title']+' - '+song_df['artist_name']
song_df

user_id             song_id  \
0        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOAKIMP12A8C130995   
1        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBBMDR12A8C13253B   
2        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBXHDL12A81C204C0   
3        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBYHAJ12A6701BF1D   
4        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SODACBL12A8C13C273   
...                                           ...                 ...   
1999995  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJEYPO12AAA8C6B0E   
1999996  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJJYDE12AF729FC16   
1999997  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJKQSF12A6D4F5EE9   
1999998  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJUXGA12AC961885C   
1999999  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJYOLS12A8C13C06F   

         listen_count                           title  \
0                   1                        The Cove   
1                   2                 Entre Dos Aguas   
2                   1                        Stronger   
3                   1                  Constellations   
4                   1                    Learn To Fly   
...               ...                             ...   
1999995             2       Ignorance (Album Version)   
1999996             4          Two Is Better Than One   
1999997             3  What I've Done (Album Version)   
1999998             1                              Up   
1999999             1      Soil_ Soil (Album Version)   

                               release  \
0                   Thicker Than Water   
1                  Flamenco Para Niños   
2                           Graduation   
3                    In Between Dreams   
4        There Is Nothing Left To Lose   
...                                ...   
1999995                      Ignorance   
1999996                     Love Drunk   
1999997                 What I've Done   
1999998                      My Worlds   
1999999                        The Con   

                                    artist_name  year  \
0                                  Jack Johnson     0   
1                                 Paco De Lucia  1976   
2                                    Kanye West  2007   
3                                  Jack Johnson  2005   
4                                  Foo Fighters  1999   
...                                         ...   ...   
1999995                                Paramore     0   
1999996  Boys Like Girls featuring Taylor Swift  2009   
1999997                             Linkin Park  2007   
1999998                           Justin Bieber  2010   
1999999                          Tegan And Sara  2007   

                                                      song  
0                                  The Cove - Jack Johnson  
1                          Entre Dos Aguas - Paco De Lucia  
2                                    Stronger - Kanye West  
3                            Constellations - Jack Johnson  
4                              Learn To Fly - Foo Fighters  
...                                                    ...  
1999995               Ignorance (Album Version) - Paramore  
1999996  Two Is Better Than One - Boys Like Girls featu...  
1999997       What I've Done (Album Version) - Linkin Park  
1999998                                 Up - Justin Bieber  
1999999        Soil_ Soil (Album Version) - Tegan And Sara  

[2000000 rows x 8 columns]

In [ ]:
song_id_grouped = song_df.groupby(['song_id', 'song']).agg({'listen_count':'count'}).reset_index()
song_id_grouped.sort_values(['listen_count', 'song_id'], ascending=[0,1])

song_id                                               song  \
2220  SOFRQTD12A81C233C0                           Sehr kosmisch - Harmonia   
317   SOAUWYT12A81C206F1                                       Undo - Björk   
352   SOAXGDH12A8C13F8A1  Dog Days Are Over (Radio Edit) - Florence + Th...   
614   SOBONKR12A58A7A7E0                     You're The One - Dwight Yoakam   
7416  SOSXLTC12AF72A7F54                            Revelry - Kings Of Leon   
...                  ...                                                ...   
4492  SOLIGVL12AB017DBAE              Historia Del Portero - Ricardo Arjona   
8747  SOWNLZF12A58A79811                          Scared - Three Days Grace   
622   SOBPGWB12A6D4F7EF3                    Don´t Leave Me Now - Amparanoia   
2666  SOGSPGJ12A8C134FAA              Ghosts (Toxic Avenger Mix) - Ladytron   
9638  SOYYBJJ12AB017E9FD                       No Creo En El Jamas - Juanes   

      listen_count  
2220          8277  
317           7032  
352           6949  
614           6412  
7416          6145  
...            ...  
4492            51  
8747            51  
622             50  
2666            48  
9638            48  

[10000 rows x 3 columns]

In [ ]:
#Agrupar por titulo
title_grouped = song_df.groupby(['title']).agg({'listen_count':'count'}).reset_index()
title_grouped.sort_values(['listen_count', 'title'], ascending=[0,1])[0:9]

title  listen_count
6836                                      Sehr kosmisch          8277
8725                                               Undo          7032
1964                     Dog Days Are Over (Radio Edit)          6949
9496                                     You're The One          6729
6498                                            Revelry          6145
6825                                            Secrets          5841
3437  Horn Concerto No. 4 in E flat K495: II. Romanc...          5385
2595                                          Fireflies          4795
3322                                   Hey_ Soul Sister          4758

In [ ]:
#Agrupar por canción
song_grouped = song_df.groupby(['song']).agg({'listen_count':'count'}).reset_index()
song_grouped

song  listen_count
0     #!*@ You Tonight [Featuring R. Kelly] (Explici...            78
1                              #40 - DAVE MATTHEWS BAND           338
2                                   & Down - Boys Noize           373
3                             ' Cello Song - Nick Drake           103
4                           '97 Bonnie & Clyde - Eminem            93
...                                                 ...           ...
9948                               the Love Song - K-OS            70
9949                  you were there with me - Four Tet            79
9950       ¡Viva La Gloria! (Album Version) - Green Day           171
9951            ¿Lo Ves? [Piano Y Voz] - Alejandro Sanz            71
9952                              Época - Gotan Project           182

[9953 rows x 2 columns]

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
sns.countplot(song_df['song'])
fig = plt.gcf()
fig.set_size_inches(10,10)
plt.title('Type')

ValueError: ignored

In [ ]:
song_grouped.sort_values(['listen_count', 'song'], ascending=[0,1])[0:9]

song  listen_count
7127                           Sehr kosmisch - Harmonia          8277
9084                                       Undo - Björk          7032
2068  Dog Days Are Over (Radio Edit) - Florence + Th...          6949
9880                     You're The One - Dwight Yoakam          6412
6774                            Revelry - Kings Of Leon          6145
7115                              Secrets - OneRepublic          5841
3613  Horn Concerto No. 4 in E flat K495: II. Romanc...          5385
2717                     Fireflies - Charttraxx Karaoke          4795
3485                           Hey_ Soul Sister - Train          4758

In [ ]:
#Agrupar por artista
artist_grouped = song_df.groupby(['artist_name']).agg({'listen_count':'count'}).reset_index()
artist_grouped.sort_values(['listen_count', 'artist_name'], ascending=[0,1])[0:9]

artist_name  listen_count
649                 Coldplay         29422
2850          The Black Keys         19862
1651           Kings Of Leon         18747
1107  Florence + The Machine         18112
1370            Jack Johnson         17801
2946             The Killers         16063
2374               Radiohead         14890
736                Daft Punk         14715
2073                    Muse         14005

In [ ]:
#Agrupar por año
year_grouped = song_df.groupby(['year']).agg({'listen_count':'count'}).reset_index()
year_grouped.sort_values(['listen_count', 'year'], ascending=[0,1])[0:9]

year  listen_count
0      0        372122
56  2009        179168
55  2008        166788
54  2007        162573
53  2006        123331
52  2005        113996
50  2003        105226
51  2004         73845
49  2002         71825

In [ ]:
year_grouped.sort_values(['listen_count', 'year'], ascending=[0,1])[1:10]

year  listen_count
56  2009        179168
55  2008        166788
54  2007        162573
53  2006        123331
52  2005        113996
50  2003        105226
51  2004         73845
49  2002         71825
57  2010         66914

In [ ]:
year_grouped.sort_values(['listen_count', 'year'], ascending=[0,1])[0:9]

year  listen_count
0      0        372122
56  2009        179168
55  2008        166788
54  2007        162573
53  2006        123331
52  2005        113996
50  2003        105226
51  2004         73845
49  2002         71825

In [ ]:
#Numero de canciones unicas
len(song_df.song.unique())

9953

In [ ]:
len(song_df.song_id.unique())

10000

In [ ]:
#Numero de usuarios totales
len(song_df.user_id.unique())

76353

In [ ]:
#Calculamos cuantos canciones ha escuchado cada persona
song_user = song_df.groupby('user_id')['song'].count()
song_user




user_id
00003a4459f33b92906be11abe0e93efc423c0ff     7
00005c6177188f12fb5e2e82cdbd93e8a3f35e64     5
00030033e3a2f904a48ec1dd53019c9969b6ef1f     9
0007235c769e610e3d339a17818a5708e41008d9    10
0007c0e74728ca9ef0fe4eb7f75732e8026a278b     9
                                            ..
fffce9c1537fbc350ea68823d956eaa8f5236dbe    44
fffd6a2bdef646ce9898b628d5dd56c43df69a9d    11
fffd9635b33f412de8ed02e44e6564e3644cf3c6    17
fffe6d1d8500f1c1f31bd63abce35c0f975a86bf     7
fffea3d509760c984e7d40789804c0e5e289cc86    23
Name: song, Length: 76353, dtype: int64

In [ ]:
song_user.median()

16.0

In [ ]:
# Get users which have listen to at least 16 songs
song_ten_id = song_user[song_user > 16].index.to_list()

In [ ]:
# Filtered the dataset to keep only those users with more than 16 listened
song_df_more_ten = song_df[song_df['user_id'].isin(song_ten_id)].reset_index(drop=True)
song_df_more_ten

user_id             song_id  \
0        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOAKIMP12A8C130995   
1        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBBMDR12A8C13253B   
2        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBXHDL12A81C204C0   
3        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBYHAJ12A6701BF1D   
4        b80344d063b5ccb3212f76538f3d9e43d87dca9e  SODACBL12A8C13C273   
...                                           ...                 ...   
1635297  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJEYPO12AAA8C6B0E   
1635298  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJJYDE12AF729FC16   
1635299  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJKQSF12A6D4F5EE9   
1635300  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJUXGA12AC961885C   
1635301  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJYOLS12A8C13C06F   

         listen_count                           title  \
0                   1                        The Cove   
1                   2                 Entre Dos Aguas   
2                   1                        Stronger   
3                   1                  Constellations   
4                   1                    Learn To Fly   
...               ...                             ...   
1635297             2       Ignorance (Album Version)   
1635298             4          Two Is Better Than One   
1635299             3  What I've Done (Album Version)   
1635300             1                              Up   
1635301             1      Soil_ Soil (Album Version)   

                               release  \
0                   Thicker Than Water   
1                  Flamenco Para Niños   
2                           Graduation   
3                    In Between Dreams   
4        There Is Nothing Left To Lose   
...                                ...   
1635297                      Ignorance   
1635298                     Love Drunk   
1635299                 What I've Done   
1635300                      My Worlds   
1635301                        The Con   

                                    artist_name  year  \
0                                  Jack Johnson     0   
1                                 Paco De Lucia  1976   
2                                    Kanye West  2007   
3                                  Jack Johnson  2005   
4                                  Foo Fighters  1999   
...                                         ...   ...   
1635297                                Paramore     0   
1635298  Boys Like Girls featuring Taylor Swift  2009   
1635299                             Linkin Park  2007   
1635300                           Justin Bieber  2010   
1635301                          Tegan And Sara  2007   

                                                      song  
0                                  The Cove - Jack Johnson  
1                          Entre Dos Aguas - Paco De Lucia  
2                                    Stronger - Kanye West  
3                            Constellations - Jack Johnson  
4                              Learn To Fly - Foo Fighters  
...                                                    ...  
1635297               Ignorance (Album Version) - Paramore  
1635298  Two Is Better Than One - Boys Like Girls featu...  
1635299       What I've Done (Album Version) - Linkin Park  
1635300                                 Up - Justin Bieber  
1635301        Soil_ Soil (Album Version) - Tegan And Sara  

[1635302 rows x 8 columns]

In [ ]:

len(song_df_more_ten.user_id.unique())

36561

In [ ]:
len(song_df_more_ten.song.unique())

9953

In [ ]:
len(song_df_more_ten.song_id.unique())

10000

In [ ]:
# convert the dataframe into a pivot table
df_songs_features = song_df_more_ten.pivot_table(index='song', columns='user_id', values='listen_count', aggfunc='sum').fillna(0)



In [ ]:
from scipy.sparse import csr_matrix

In [ ]:
# obtain a sparse matrix
mat_songs_features = csr_matrix(df_songs_features.values)

In [ ]:
#Agrupar por canción
song_grouped = song_df_more_ten.groupby('song')['user_id'].count()
song_grouped


song
#!*@ You Tonight [Featuring R. Kelly] (Explicit Album Version) - The Notorious B.I.G.     67
#40 - DAVE MATTHEWS BAND                                                                 272
& Down - Boys Noize                                                                      343
' Cello Song - Nick Drake                                                                 87
'97 Bonnie & Clyde - Eminem                                                               71
                                                                                        ... 
the Love Song - K-OS                                                                      51
you were there with me - Four Tet                                                         69
¡Viva La Gloria! (Album Version) - Green Day                                             135
¿Lo Ves? [Piano Y Voz] - Alejandro Sanz                                                   59
Época - Gotan Project                                            

In [ ]:
# Get songs which have been listened at least 4 times
user_ten_id  = song_grouped[song_grouped > 200].index.to_list()

In [ ]:
df_song_reduced = song_df_more_ten[(song_df_more_ten['song'].isin(user_ten_id))].reset_index(drop=True)
df_song_reduced

user_id             song_id  \
0       b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBXHDL12A81C204C0   
1       b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBYHAJ12A6701BF1D   
2       b80344d063b5ccb3212f76538f3d9e43d87dca9e  SODACBL12A8C13C273   
3       b80344d063b5ccb3212f76538f3d9e43d87dca9e  SODXRTY12AB0180F3B   
4       b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOFRQTD12A81C233C0   
...                                          ...                 ...   
841582  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOIOESO12A6D4F621D   
841583  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJEYPO12AAA8C6B0E   
841584  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJJYDE12AF729FC16   
841585  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJKQSF12A6D4F5EE9   
841586  d8bfd4ec88f0f3773a9e022e3c1a0f1d3b7b6a92  SOJUXGA12AC961885C   

        listen_count                           title  \
0                  1                        Stronger   
1                  1                  Constellations   
2                  1                    Learn To Fly   
3                  1                   Paper Gangsta   
4                  1                   Sehr kosmisch   
...              ...                             ...   
841582             4          Unwell (Album Version)   
841583             2       Ignorance (Album Version)   
841584             4          Two Is Better Than One   
841585             3  What I've Done (Album Version)   
841586             1                              Up   

                              release                             artist_name  \
0                          Graduation                              Kanye West   
1                   In Between Dreams                            Jack Johnson   
2       There Is Nothing Left To Lose                            Foo Fighters   
3                    The Fame Monster                               Lady GaGa   
4                  Musik von Harmonia                                Harmonia   
...                               ...                                     ...   
841582    More Than You Think You Are                         matchbox twenty   
841583                      Ignorance                                Paramore   
841584                     Love Drunk  Boys Like Girls featuring Taylor Swift   
841585                 What I've Done                             Linkin Park   
841586                      My Worlds                           Justin Bieber   

        year                                               song  
0       2007                              Stronger - Kanye West  
1       2005                      Constellations - Jack Johnson  
2       1999                        Learn To Fly - Foo Fighters  
3       2008                          Paper Gangsta - Lady GaGa  
4          0                           Sehr kosmisch - Harmonia  
...      ...                                                ...  
841582  2003           Unwell (Album Version) - matchbox twenty  
841583     0               Ignorance (Album Version) - Paramore  
841584  2009  Two Is Better Than One - Boys Like Girls featu...  
841585  2007       What I've Done (Album Version) - Linkin Park  
841586  2010                                 Up - Justin Bieber  

[841587 rows x 8 columns]

In [ ]:
x = df_song_reduced.groupby(['song', 'user_id']).agg({'listen_count':'sum'}).reset_index()
x.sort_values(['listen_count'], ascending=[0])[0:19]

song  \
102852       Call It Off (Album Version) - Tegan And Sara   
833297                     You're The One - Dwight Yoakam   
338915         I'm On A Boat - The Lonely Island / T-Pain   
146601                        Dance_ Dance - Fall Out Boy   
550712                  Reelin' In The Years - Steely Dan   
209762              Fast Car (LP Version) - Tracy Chapman   
526164                          Peace Train - Cat Stevens   
248900                      Glamorous - Fergie / Ludacris   
558997                            Revelry - Kings Of Leon   
7685    A Beggar On A Beach Of Gold - Mike And The Mec...   
831794                     You're The One - Dwight Yoakam   
686346                     The Gift - Angels and Airwaves   
484718  Nine Times Out Of Ten (1998 Digital Remaster) ...   
66026                  Bird Song - Florence + The Machine   
548866   Recado Falado (Metrô Da Saudade) - Alceu Valença   
333847                  I'll Be Your Man - The Black Keys   
560668                            Revelry - Kings Of Leon   
678997                      The Big Gundown - The Prodigy   
560820                            Revelry - Kings Of Leon   

                                         user_id  listen_count  
102852  5ea608df0357ec4fda191cb9316fe8e6e65e3777           879  
833297  c012ec364329bb08cbe3e62fe76db31f8c5d8ec3           683  
338915  70caceccaa745b6f7bc2898a154538eb1ada4d5a           676  
146601  972cce803aa7beceaa7d0039e4c7c0ff097e4d55           664  
550712  d2232ac7a1ec17b283b5dff243161902b2cb706c           649  
209762  70caceccaa745b6f7bc2898a154538eb1ada4d5a           553  
526164  2be4e2736f580dd4fe7b489f675935ccfad0a453           500  
248900  c012ec364329bb08cbe3e62fe76db31f8c5d8ec3           500  
558997  7d3cb0aca2aff23e5fd06e34d66ec16ccfdaab4e           488  
7685    d4438263e80ea31cbcca30005873a4074a0c248e           453  
831794  66ff466e107e9bcd7c6f2c59501ed0d7844c26c0           446  
686346  6040ed6d4e10f09e6cccd64c819c684531abe682           436  
484718  e9d8dfd545ca342caa23202ceb2d969ce04dc030           431  
66026   dc20d1116f083d39d3035e13e266b45d72f9a627           419  
548866  0ee4cadeda8f05272083df630fdc8986afd47e9e           413  
333847  09318175bf2f39d316f41d80fbfc4e4093cf0a19           405  
560668  df3aec7a12200e9f2479c0540bb7322c27e7c974           372  
678997  87254d6c37f86b17ebc685f563ea71dfa1b4966f           372  
560820  e8b6056b402ea930b4e40c47ee8c2c1083b74739           353

In [ ]:
# Determine the bins. If a user listen to a song more than 10 times, 
# we will assign 10
bins = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9,  2214]

# Binning
df_song_reduced['listen_count'] = pd.cut(df_song_reduced['listen_count'], bins=bins, labels=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

In [ ]:

df_song_reduced.groupby('listen_count')['listen_count'].count()

listen_count
1     458679
2     137637
3      65060
4      38135
5      37831
6      21120
7      14388
8      10409
9       7773
10     50555
Name: listen_count, dtype: int64